sql/core/src/test/scala/org/apache/spark/sql/ArrowConvertersSuite.scala

```diff
@@ -0,0 +1,568 @@
+      .toDF("upper_case", "lower_case", "null_str")
+    DataTuple(df, new JSONFile(schema, Seq(batch)), "stringData.json")
+  }
+
+  private def byteData: DataTuple = {
+    val data = List[Byte](1.toByte, (-1).toByte, 64.toByte, Byte.MaxValue)
+    val fields = Seq(new IntegerType("a_byte", is_signed = true, 8, nullable = false))
+    val schema = new JSONSchema(fields)
+    val columns = Seq(
+      new PrimitiveColumn("a_byte", data.length, data.map(_ => true), data.map(_.toInt)))
+    val batch = new JSONRecordBatch(data.length, columns)
+    DataTuple(data.toDF("a_byte"), new JSONFile(schema, Seq(batch)), "byteData.json")
+  }
+
+  private def floatNaNData: DataTuple = {
+    val fnan = Seq(1.2F, Float.NaN)
+    val dnan = Seq(Double.NaN, 1.2)
+    val fields = Seq(new FloatingPointType("NaN_f", 32, nullable = false),
+      new FloatingPointType("NaN_d", 64, nullable = false))
+    val schema = new JSONSchema(fields)
+    val columns = Seq(new PrimitiveColumn("NaN_f", fnan.length, fnan.map(_ => true), fnan),
+      new PrimitiveColumn("NaN_d", dnan.length, dnan.map(_ => true), dnan))
+    val batch = new JSONRecordBatch(fnan.length, columns)
+    val df = fnan.zip(dnan).toDF("NaN_f", "NaN_d")
+    DataTuple(df, new JSONFile(schema, Seq(batch)), "nanData-floating_point.json")
+  }
+
+  private def timestampData: DataTuple = {
+    val sdf = new SimpleDateFormat("yyyy-MM-dd HH:mm:ss.SSS z", Locale.US)
+    val ts1 = new Timestamp(sdf.parse("2013-04-08 01:10:15.567 UTC").getTime)
+    val ts2 = new Timestamp(sdf.parse("2013-04-08 13:10:10.789 UTC").getTime)
+    val data = Seq(ts1, ts2)
+    val schema = new JSONSchema(Seq(new TimestampType("c_timestamp")))
+    val columns = Seq(
+      new PrimitiveColumn("c_timestamp", data.length, data.map(_ => true), data.map(_.getTime)))
+    val batch = new JSONRecordBatch(data.length, columns)
+    DataTuple(data.toDF("c_timestamp"), new JSONFile(schema, Seq(batch)), "timestampData.json")
+  }
+
+  private def dateData: DataTuple = {
+    val sdf = new SimpleDateFormat("yyyy-MM-dd HH:mm:ss.SSS z", Locale.US)
+    val d1 = new Date(sdf.parse("2015-04-08 13:10:15.000 UTC").getTime)
+    val d2 = new Date(sdf.parse("2015-04-08 13:10:15.000 UTC").getTime)
+    val ts1 = new Timestamp(sdf.parse("2013-04-08 01:10:15.567 UTC").getTime)
+    val ts2 = new Timestamp(sdf.parse("2013-04-08 13:10:10.789 UTC").getTime)
+    val df = Seq((d1, sdf.format(d1), ts1), (d2, sdf.format(d2), ts2))
+      .toDF("a_date", "b_string", "c_timestamp")
+    val jsonFile = new JSONFile(new JSONSchema(Seq.empty[DataType]), Seq.empty[JSONRecordBatch])
+    DataTuple(df, jsonFile, "dateData.json")
+  }
+
+  /**
+   * Arrow JSON Format Data Generation
+   * Referenced from https://github.com/apache/arrow/blob/master/integration/integration_test.py
+   * TODO: Look into using JSON generation from parquet-vector.jar
+   */
+
+  private abstract class DataType(name: String, nullable: Boolean) {
```

Comment:
I thought the easiest and most compact strategy for testing was to 1) define common data, 2) write the data as JSON, 3) make a DataFrame from data and convert to ArrowRecordBatch, 4) send batch and JSON file to Arrow validator where it can be checked.

I think the JSON files are too big to inline and would create a huge file if have them as strings.  I could try this out for one of the test cases if you like and we can see how it turns out?